x-pack/legacy/plugins/apm/public/components/shared/LocalFilters/Filter.tsx

```diff
@@ -0,0 +1,209 @@
+              : {})
+          } as {
+            label: string;
+            append?: React.ReactNode;
+            checked?: 'on' | 'off' | undefined;
+          })
+      ),
+    [value, options]
+  );
+
+  const [visibleOptions, setVisibleOptions] = useState(items);
+
+  const onApply = useCallback(() => {
+    const newValue = visibleOptions
+      .filter(option => option.checked === 'on')
+      .map(option => option.label);
+
+    setShowPopover(false);
+    onChange(newValue);
+  }, [onChange, visibleOptions]);
+
+  useEffect(() => {
+    setVisibleOptions(items);
+  }, [items]);
+
+  const badges = value.length ? (
+    <>
+      <EuiFlexGrid gutterSize="s">
+        {value.map(val => (
+          <EuiFlexItem key={val} grow={false}>
+            <button
+              type="button"
+              onClick={() => {
+                onChange(value.filter(v => val !== v));
+              }}
+            >
+              <EuiBadge color="hollow">
+                <BadgeText>{val}</BadgeText>
+                <EuiIcon type="cross"></EuiIcon>
+              </EuiBadge>
+            </button>
+          </EuiFlexItem>
+        ))}
+      </EuiFlexGrid>
+      <EuiSpacer size="s"></EuiSpacer>
+    </>
+  ) : (
+    []
+  );
+
+  return (
+    <>
+      <Popover
+        id={`local-filter-${name}`}
+        closePopover={toggleShowPopover}
+        button={button}
+        isOpen={showPopover}
+        panelPaddingSize="s"
+        anchorPosition="downLeft"
+      >
+        <EuiSelectable
+          onChange={selectedOptions => {
+            setVisibleOptions(selectedOptions);
+          }}
+          options={visibleOptions}
+          searchable={true}
+        >
+          {(list, search) => (
+            <SelectContainer>
+              <EuiFlexGroup direction="column" gutterSize="none">
+                <EuiFlexItem grow={true}>
+                  <EuiTitle size="xxxs" textTransform="uppercase">
+                    <h4>
+                      {i18n.translate('xpack.apm.applyFilter', {
+                        defaultMessage: 'Apply {title} filter',
+                        values: { title }
+                      })}
+                    </h4>
+                  </EuiTitle>
+                  <EuiHorizontalRule margin="none" />
+                  <EuiSpacer size="s" />
+                  {search}
+                  <EuiSpacer size="s" />
+                  {list}
+                  <EuiSpacer size="s" />
+                </EuiFlexItem>
+                <EuiFlexItem grow={false}>
+                  <EuiButton onClick={onApply} size="s">
+                    {i18n.translate('xpack.apm.applyFilter', {
+                      defaultMessage: 'Apply options',
+                      values: { title }
+                    })}
+                  </EuiButton>
+                </EuiFlexItem>
+              </EuiFlexGroup>
+            </SelectContainer>
+          )}
+        </EuiSelectable>
+      </Popover>
+      {badges}
```

Comment:
I think this component would be easier to read if some of the transformation is extracted to separate functions, and perhaps smaller components. Eg. `badges` could be a `Badge` component.

Reply:
Aye, I'm removing at least the badge list into their own file. Let's see how it looks after that.